0004-xbl-web-components.md

```diff
@@ -0,0 +1,75 @@
+
+This analysis is focused on widely used and relatively small XBL components that could potentially be swapped out more-or-less in place with corresponding Web Components. For instance: buttons, labels, input fields, etc.
+
+Components that are only used once or are implemented mostly in C++ are ignored in this analysis. For instance: tabbrowser, trees, etc. Most of the same principles that apply to smaller components should also apply to them, but we may also consider different options for moving them away from XBL.
+
+## Auditing the XBL feature set
+
+From the [Introduction to XBL](https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XUL/Tutorial/Introduction_to_XBL), a binding has five types of things that it declares: Content, properties, methods, events, and style.  But it's helpful to look at specific features used in Firefox and comparing how they map back to Web Components.
+
+Which bindings use which features is visualized at https://bgrins.github.io/xbl-analysis/. The scripts to generate that tree and also to semi-automate translation of bindings can be found at https://github.com/bgrins/xbl-analysis.
+
+### XBL features that map to Web Components features
+
+1. `<property>`: These map to JS getters and setters on the Custom Element class (‘val’ name is magic in xbl, would need to keep same name to translate directly)
+2. `<method>`: These map to functions on the Custom Element class
+3. `<constructor>` This maps to connectedCallback on the Custom Element class (which is called when the element is inserted into a document).
+In XBL, the constructor fires after the element is inserted into the document (since the binding is attached through a CSS property).  However, the constructor will wait some additional time to fire, unless if the element is accessed via JS in which case it will run immediately
+4. `<destructor>`: These map to `disconnectedCallback` on the Custom Element class
+5. `[extends]`: A way to extend the functionality of another binding, i.e. `extends="chrome://global/content/bindings/general.xml#basetext"`. On a surface level this appears to map directly to how the ‘extends’ keyword in ES6 classes behavior with Custom Elements, but more investigation is needed.  There’s a small set of tags where you can do `extends="xul:button"` but that’s aliasing the same thing.
+6. `<handlers>`: Can be implemented with addEventListener or possibly a helper on the base element i.e:
+```
+helpers: {
+  "click": () => { /* bubble by default*/ }
+  "focus|capturing": () => { /* copy in xbl implementation */ }
+}
+```
+7. `[inherits]`: This is a way to declaratively copy attributes from a binding parent to content and keep them in sync, similar to the `<observes>` element. For example, this copies the feedURL from the parent into the 'value' field of the label, and also observes future changes
```

Comment:
Yes, it can be done with a MutationObserver and a bit of glue. This is prototyped out here: https://github.com/bgrins/xbl-analysis/blob/gh-pages/elements/base-element.js#L29.

I'll add some notes about this in the doc.